### DISCUSSION TITLE
How to share a common state between web app and android (Presence?)

### DISCUSSION THREAD

**Discussion by boulaycote** [supabase-community/supabase-kt#130] [Q&A] [answered]
Hi,

I want to have a common state shared across devices. In this current case, between a web app and an android app. I have tried using `Presence` and `channel.track(state)` but the state, when sending from android, seems to be filtered out and the only thing being received on the web app end is the `presence_ref` field. 

First, is this a good use case for `Presence`? And if so, any idea what's happening?

Thank you

**Comment by jan-tennert** [comment]
Not sure what you mean with "state", can you describe it a little bit more? Can you show me how you receive these presences? Also make sure both channels have the same id

**Reply by boulaycote** [reply]
Let's say I have an object describing a state like so: 

```
val state = {
    currentPage: 'menu',
    selectedProduct: '12345'
}    
```

I am sending this state through `channel?.track(state)` on the android side. Then on the web app side I am receiving a `presence sync` event

```
channel.on('presence', { event: 'sync' }, () => {
    const presence = channel.presenceState()
    // update local state [...]
}
```
Both channels have the same Id, I am already sending broadcasts and actually am receiving the `presence sync` events but without the `currentPage` and `selectedProduct` keys.

**Reply by jan-tennert** [reply]
Okay so assuming you have the same channel id for both clients, can you try to enable [logging](https://github.com/supabase-community/supabase-kt/wiki/Logging) and see what is says?

**Reply by boulaycote** [reply]
```
meViewModel$testMessage  D  Sending PresenceState(test=2, selectedProduct=test, currentPage=test)
ltimeImpl$sendHeartbeat  D  Sending heartbeat
 RealtimeImpl$onMessage  D  Received message {"event":"phx_reply","payload":{"response":{},"status":"ok"},"ref":"23","topic":"realtime:12345"}
                         D  Received event phx_reply for channel realtime:12345
                         D  Received message {"event":"presence_diff","payload":{"joins":{"12345":{"metas":[{"phx_ref":"F1tXOf0DbpQjpaAm","phx_ref_prev":"F1tXONbNnQEjpZGm"}]}},"leaves":{"12345":{"metas":[{"phx_ref":"F1tXONbNnQEjpZGm","phx_ref_prev":"F1tXOL5hLdQjpUZB"}]}}},"ref":null,"topic":"realtime:12345"}
                         D  Received event presence_diff for channel realtime:12345
                         D  Received message {"event":"phx_reply","payload":{"response":{},"status":"ok"},"ref":"24","topic":"phoenix"}
                         I  Heartbeat received
```

and on the web app side 

```
Prensence sync {"12345":[{"presence_ref":"F1tXOf0DbpQjpaAm"}]}
```

**Reply by jan-tennert** [reply]
Okay I may have identified the problem, I'm publishing a dev version for you to test out, give me a few minutes

**Reply by jan-tennert** [reply, marked as answer]
Can you try out version `0.9.5-dev-1` and tell me if it works now? (Edit: fixed in 0.9.5)

**Reply by boulaycote** [reply]
Yes this appears to be working!

And now, for the second question: do you think this is a good use case for presence? Maintaining a shared state? If so, I think I have to merge the different presence(s) into one, as the object looks like:

```
{"12345":[
    {"currentPage":"current-page","currentProductId":"current-product-id","selectedProductId":"selected-product-id","presence_ref":"F1tcQ85B8yqM2qeI"},
    {"currentPage":"current-page","presence_ref":"F1tcTB6pTmkjpUrm"}
]}
```

In my case I can assume only the relevant clients are connected to the channel (id: 12345).

**Reply by jan-tennert** [reply]
I see what you mean. Using Presence maybe doesn't fit that well as each client still has an own state and as you said you kinda have to merge them together on each change. Another possibility would be to have a table holding a state for every user and each client listening for changes on a single row. (If you have questions regarding this, feel free to ask me)
I'm not sure, you can try asking in the supabase discord server or in their github discussion for further help. 
But I'm glad this was fixed.

**Reply by boulaycote** [reply]
Yes I guess a table would work. I'll figure it out thanks! I would still use presence for its intended use case, to know if clients are connected.

**Reply by boulaycote** [reply]
And using broadcast would work as well right? Just not guaranteed delivery?

**Reply by jan-tennert** [reply]
> And using broadcast would work as well right? Just not guaranteed delivery?

Broadcast is just like sending a message directly to all clients connected to a channel. So as long all are connected, they should get it

**Reply by jan-tennert** [reply]
> Yes I guess a table would work. I'll figure it out thanks! I would still use presence for its intended use case, to know if clients are connected.

Let me know if you need any assistance!
